File: DataStructures/week2_priority_queues_and_disjoint_sets/3_merging_tables/3_merging_tables.py
#!/usr/bin/env python
# coding: utf-8

# In[ ]:


# python3
"""
In this problem, your goal is to simulate a sequence of merge operations with tables in a
database.
"""

class Database:
    def __init__(self, row_counts):
        self.row_counts = row_counts
        self.max = max(self.row_counts)
        n_tables = len(row_counts)
        self.ranks = [1] * (n_tables)
        self.parents = list(range(0, n_tables))

    def merge(self, src, dst):
        src_parent = self.get_parent(src)
        dst_parent = self.get_parent(dst)

        if src_parent == dst_parent:
            return

        # merge two components
        # use union by rank heuristic
        # update max_row_count with the new maximum table size
        if self.ranks[src_parent] > self.ranks[dst_parent]:
            self.parents[dst_parent] = src_parent
        else:
            self.parents[src_parent] = dst_parent
            if self.ranks[src_parent] == self.ranks[dst_parent]:
                self.ranks[src_parent] += 1

        self.row_counts[dst_parent] += self.row_counts[src_parent]
        self.row_counts[src_parent] = 0

        if self.max < self.row_counts[dst_parent]:
            self.max = self.row_counts[dst_parent]
        return True

    def get_parent(self, table):
        # here, table is the index where the table belongs
        update_parents = []
        
#         root = table
        while table != self.parents[table]:
            update_parents.append(self.parents[table])
            table = self.parents[table]
        
        for i in update_parents:
            self.parents[i] = table
            
        return table


def main():
    n_tables, n_queries = map(int, input().split())
    counts = list(map(int, input().split()))
    assert len(counts) == n_tables
    db = Database(counts)
    for i in range(n_queries):
        dst, src = map(int, input().split())
        db.merge(dst - 1, src - 1)
        print(db.max)


if __name__ == "__main__":
    main()

File: DataStructures/week2_priority_queues_and_disjoint_sets/3_merging_tables/3_merging_tables.cpp
#include <cstdio>
#include <cstdlib>
#include <vector>
#include <algorithm>
#include <iostream>

using std::cin;
using std::cout;
using std::endl;
using std::max;
using std::vector;
/*
In this problem, your goal is to simulate a sequence of merge operations with tables in a
database.
*/

struct DisjointSets {
	vector<int> row_counts;
	int max_table_size;
	vector <int> parents;
	vector<int> ranks;

	DisjointSets(vector<int> row_counts) : 
		row_counts(row_counts), max_table_size(*max_element(row_counts.begin(), row_counts.end())){
		for (int i = 0; i < row_counts.size(); i++) {
			parents.push_back(i);
			ranks.push_back(1);
		}
	}

	int getParent(int table) {
		// find parent and compress path
		vector<int> update_parents;

		while (table != parents[table]) {
			update_parents.push_back(parents[table]);
			table = parents[table];
		}

		for (auto elem : update_parents) parents[elem] = table;

		return table;
	}

	void merge(int destination, int source) {
		int realDestination = getParent(destination);
		int realSource = getParent(source);
		if (realDestination != realSource) {
			// merge two components
			// use union by rank heuristic
						// update max_table_size
			if (ranks[realSource] > ranks[realDestination]) {
				parents[realDestination] = realSource;
			}
			else {
				parents[realSource] = realDestination;
				if (ranks[realSource] == ranks[realDestination])
					ranks[realSource] += 1;
			}
			row_counts[realDestination] += row_counts[realSource];
			row_counts[realSource] = 0;

			if (max_table_size < row_counts[realDestination])
				max_table_size = row_counts[realDestination];
		}
	}
};

int main() {
	int n, m;
	cin >> n >> m;

	vector<int> counts(n);
	for (int i = 0; i != n; ++i) {
		cin >> counts[i];
	}

	DisjointSets tables(counts);

	for (int i = 0; i != m; ++i) {
		int destination, source;
		cin >> destination >> source;
		--destination;
		--source;

		tables.merge(destination, source);
		cout << tables.max_table_size << endl;
	}

	return 0;
}
